MINOR: Add ShareFetch quota session verification test

sjhajharia opened the pull request:
### Background
As part of KIP-932 implementation, ShareFetch requests need to properly
integrate with Kafka's quota system. This requires that ShareFetch
requests extract and pass the correct session information (Principal,
client address, client ID) to quota managers, ensuring consistent quota
enforcement between ShareFetch and traditional Fetch requests.

### Changes
This PR adds `testHandleShareFetchRequestQuotaTagsVerification()`,
`testHandleShareAcknowledgeRequestQuotaTagsVerification` and
`testHandleShareFetchWithAcknowledgementQuotaTagsVerification` to
`KafkaApisTest`, which provides verification of quota tag extraction and
session handling for ShareFetch and ShareAcknowledge requests.
   - Ensures ShareFetch/ShareAck requests are properly constructed with
the correct client ID, principal, client address, and API key
   - Verifies the request context contains the expected session
information
   - Uses `ArgumentCaptor` to capture the exact `Session` and
`RequestChannel.Request` objects passed to quota managers
   - Verifies both `quotas.fetch.maybeRecordAndGetThrottleTimeMs()` and
`quotas.request.maybeRecordAndGetThrottleTimeMs()` are called with
correct parameters as and when needed.
   - Validates that the captured `RequestChannel.Request` object
maintains the correct request context information
   - Ensures the client ID passed to quota managers matches the
test-defined value
   - Verifies that in case of Acks being piggybacked on the fetch
requests, the quotas are applied only once and not twice.

Reviewers: Apoorv Mittal <apoorvmittal10@gmail.com>

sjhajharia committed:
Test ShareFetch quota tag extraction and session verification

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -9510,6 +9620,25 @@ class KafkaApisTest extends Logging {
       requestMetrics, envelope = None)
   }
 
+  private def buildRequestWithPrincipal(request: AbstractRequest,
+                                        principal: KafkaPrincipal,
+                                        clientAddress: InetAddress,
+                                        listenerName: ListenerName = ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT),
+                                        fromPrivilegedListener: Boolean = false,
+                                        requestHeader: Option[RequestHeader],
+                                        requestMetrics: RequestChannelMetrics = requestChannelMetrics): RequestChannel.Request = {
+    val buffer = request.serializeWithHeader(
+      requestHeader.getOrElse(new RequestHeader(request.apiKey, request.version, clientId, 0)))
+
+    // read the header from the buffer first so that the body can be read next from the Request constructor
+    val header = RequestHeader.parse(buffer)
+    val context = new RequestContext(header, "1", clientAddress, Optional.empty(),
+      principal, listenerName, SecurityProtocol.SSL,
+      ClientInformation.EMPTY, fromPrivilegedListener, Optional.of(kafkaPrincipalSerde))
+    new RequestChannel.Request(processor = 1, context = context, startTimeNanos = 0, MemoryPool.NONE, buffer,
+      requestMetrics, envelope = None)
+  }
```
Seems existing `buildRequest` method can be overloaded with this method. So we have common code.

apoorvmittal10 reviewed (commented):
Thanks for the test. Also should we have a similar test for `handleShareAcknowledgeRequest` as that should update the `request` quota. Also can we have a test when acknowledge data is present in ShareFetch then can we verify that request quota is not bumped twice, 1 becuase of shae fetch and other for acknowledgement.

sjhajharia committed:
Add tests

sjhajharia commented:
Thanks @apoorvmittal10 for the review. I have added a couple of tests as suggested by you - `testHandleShareAcknowledgeRequestQuotaTagsVerification` and
`testHandleShareFetchWithAcknowledgementQuotaTagsVerification` to test the two situations as you had mentioned. I have also refactored the code a bit to overload the `buildRequest()` method. 
Please re-review when possible.
Thanks!

apoorvmittal10 commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -9494,17 +9820,28 @@ class KafkaApisTest extends Logging {
                            fromPrivilegedListener: Boolean = false,
                            requestHeader: Option[RequestHeader] = None,
                            requestMetrics: RequestChannelMetrics = requestChannelMetrics): RequestChannel.Request = {
+    // DelegationTokens require the context authenticated to be non SecurityProtocol.PLAINTEXT
+    // and have a non KafkaPrincipal.ANONYMOUS principal. This test is done before the check
+    // for forwarding because after forwarding the context will have a different context.
+    // We validate the context authenticated failure case in other integration tests.
```
This should be in other method itself, as like earlier.

sjhajharia replied:
Thanks. Fixed the same.

apoorvmittal10 reviewed (approved):
LGTM, there is a minor comment, please if you can address that.

sjhajharia committed:
move comment

apoorvmittal10 commented:
Merging as unrelated test failures only in java 17, tracked as: https://issues.apache.org/jira/browse/KAFKA-19511, https://issues.apache.org/jira/browse/KAFKA-19513
